File: grapple-frontend/src/api-requests/promotion.ts
import { promotionsApi } from "@/hook/base-apis";

export const ADULT_BELTS = ['white', 'blue', 'purple', 'brown', 'black', 'coral', 'red/white', 'red'] as const;
export const KIDS_BELTS = [
  'white',
  'grey/white', 'grey', 'grey/black',
  'yellow/white', 'yellow', 'yellow/black',
  'orange/white', 'orange', 'orange/black',
  'green/white', 'green', 'green/black',
] as const;

export type AdultBelt = typeof ADULT_BELTS[number];
export type KidsBelt = typeof KIDS_BELTS[number];
export type BeltSystem = 'adult' | 'kids';

export interface Promotion {
  id?: string;
  gym_id: string;
  member_id: string;
  member_name: string;
  avatar_url?: string;
  system: BeltSystem;
  belt: string;
  stripes: number;
  notes?: string;
  promoted_by?: string;
  promoted_at: string;
  created_at?: string;
}

const authHeaders = (token: string) => ({
  headers: { Authorization: `Bearer ${token}` },
});

export const getPromotionHistory = async (gymId: string, memberId: string, token: string): Promise<Promotion[]> => {
  const { data } = await promotionsApi.get<Promotion[]>('', {
    params: { gym_id: gymId, member_id: memberId },
    ...authHeaders(token),
  });
  return data;
};

export const getCurrentBelts = async (gymId: string, token: string): Promise<Record<string, Promotion>> => {
  const { data } = await promotionsApi.get<Record<string, Promotion>>('/current', {
    params: { gym_id: gymId },
    ...authHeaders(token),
  });
  return data;
};

export const recordPromotion = async (
  payload: Omit<Promotion, 'id' | 'created_at'>,
  token: string
): Promise<Promotion> => {
  const { data } = await promotionsApi.post<Promotion>('', payload, authHeaders(token));
  return data;
};

export const deletePromotion = async (gymId: string, promotionId: string, token: string): Promise<void> => {
  await promotionsApi.delete(`/${gymId}/${promotionId}`, authHeaders(token));
};

File: backend/internal/dao/promotion.go
package dao

import (
	"fmt"
	"time"

	"go.mongodb.org/mongo-driver/v2/bson"
)

const (
	SystemAdult = "adult"
	SystemKids  = "kids"
)

// Adult BJJ belt progression
var AdultBelts = []string{
	"white", "blue", "purple", "brown", "black",
	"coral", "red/white", "red",
}

// Kids BJJ belt progression (IBJJF standard)
var KidsBelts = []string{
	"white",
	"grey/white", "grey", "grey/black",
	"yellow/white", "yellow", "yellow/black",
	"orange/white", "orange", "orange/black",
	"green/white", "green", "green/black",
}

// Promotion records a single belt/stripe change for a gym member.
// The member's current belt is always derived from their most recent Promotion —
// there is no separate "current_belt" field to keep in sync.
type Promotion struct {
	ID         bson.ObjectID `json:"id,omitempty" bson:"_id,omitempty"`
	GymID      bson.ObjectID `json:"gym_id" bson:"gym_id"`
	MemberID   string        `json:"member_id" bson:"member_id"`
	MemberName string        `json:"member_name" bson:"member_name"`
	AvatarURL  string        `json:"avatar_url,omitempty" bson:"avatar_url,omitempty"`
	System     string        `json:"system" bson:"system"`       // "adult" | "kids"
	Belt       string        `json:"belt" bson:"belt"`           // e.g. "blue", "grey/black"
	Stripes    int           `json:"stripes" bson:"stripes"`     // 0–4
	Notes      string        `json:"notes,omitempty" bson:"notes,omitempty"`
	PromotedBy string        `json:"promoted_by,omitempty" bson:"promoted_by,omitempty"`
	PromotedAt time.Time     `json:"promoted_at" bson:"promoted_at"`
	CreatedAt  time.Time     `json:"created_at" bson:"created_at"`
}

// ValidateBelt returns an error if belt is not valid for the given system.
func ValidateBelt(system, belt string) error {
	var valid []string
	switch system {
	case SystemAdult:
		valid = AdultBelts
	case SystemKids:
		valid = KidsBelts
	default:
		return fmt.Errorf("system must be %q or %q", SystemAdult, SystemKids)
	}
	for _, b := range valid {
		if b == belt {
			return nil
		}
	}
	return fmt.Errorf("belt %q is not valid for system %q", belt, system)
}

// ValidateStripes returns an error if stripes is out of range.
func ValidateStripes(stripes int) error {
	if stripes < 0 || stripes > 4 {
		return fmt.Errorf("stripes must be 0–4, got %d", stripes)
	}
	return nil
}
